[MLIR][NVGPU] Introduce `warpgroup.init.accumulator` Op

grypp opened the pull request:
This Op generates and initilizes the accumulator matrix for `nvgpu.warpgroup.mma` op to perform matrix-multiply-and-accumulate (mma).

Its associated transformation generates `!llvm.struct<>` and fill it with the initial values. The size of struct is number of required inout registers for `nvgpu.warpgroup.mma` op.

qcolombet commented on mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td:
```diff
@@ -727,4 +727,15 @@ def NVGPU_WarpgroupMmaOp : NVGPU_Op<"warpgroup.mma"> {
+  let summary = "Initialize accumulator matrix for `warppgroup.mma`";
+
+  let description = [{
+    This Op generates and initilizes the accumulator matrix for 
```
Nit: Initi**a**lizes

qcolombet commented on mlir/lib/Conversion/NVGPUToNVVM/NVGPUToNVVM.cpp:
```diff
@@ -1306,6 +1306,35 @@ struct NVGPUWarpgroupMmaOpLowering
+                  ConversionPatternRewriter &rewriter) const override {
+    Location loc = op->getLoc();
+    SmallVector<Value> results;
+    for (auto matrixD : op.getMatrixC()) {
```
Nit: Use the explicit type instead of `auto` here

qcolombet commented on mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td:
```diff
@@ -727,4 +727,15 @@ def NVGPU_WarpgroupMmaOp : NVGPU_Op<"warpgroup.mma"> {
+    This Op generates and initilizes the accumulator matrix for 
+    `nvgpu.warpgroup.mma` op to perform matrix-multiply-and-accumulate (mma).
+  }];
+  let results = (outs Variadic<NVGPU_WarpgroupAccumulator>:$matrixC);
```
What's the use case for the variadic part of this instruction as opposed to using several `warpgroup.mma.init.accumulator`?

grypp replied:
Makes storing them to memref easier. Otherwise, I need to write store and subview for each of accumulator type.

qcolombet replied:
> Makes storing them to memref easier. For example, multiple accumulator (with variadic) can be stored to the same memref, and their offset is nicely calculated.

But that's on the `store` side, for the `init` this doesn't make things easier (well aside from making the IR denser :)). We still create one struct per result.

> Otherwise, I need to write store for each accumulator, calculate memref's subview for each of them

We still have to do that, don't we?

grypp replied:
You are right there are trade-offs. I have 3 versions in my mind, let me put here so we can discuss. 

1) Current version with the variadic is below. All Ops has variadic, we genereate single Op for each of them.
```
// Init
%matrixC1, %matrixC2 = nvgpu.wargroup.mma.init.accumulator ->  
                    !nvgpu.wargroup.accumulator<fragmented = vector<64x128xf32>>,
                    !nvgpu.wargroup.accumulator<fragmented = vector<64x128xf32>>

// GEMM
%matrixD1, %matrixD2 = nvgpu.wargroup.mma %descA, %descB, %matrixC1, %matrixC2 ...

// Epilogue 
nvgpu.wargroup.mma.store [%matrixD1, %matrixD2] to %sharedMemoryBuffer
  : !nvgpu.wargroup.accumulator<fragmented = vector<64x128xf32>>, 
    !nvgpu.wargroup.accumulator<fragmented = vector<64x128xf32>>
    into memref<128x128xf32,3>
```

2) We can get rid of variadic in `nvgpu.wargroup.mma.init.accumulator`, like below:
```
// Init
%matrixC1 = nvgpu.wargroup.mma.init.accumulator ->  !nvgpu.wargroup.accumulator<fragmented = vector<64x128xf32>>
%matrixC2 = nvgpu.wargroup.mma.init.accumulator ->  !nvgpu.wargroup.accumulator<fragmented = vector<64x128xf32>>                    

// GEMM
%matrixD1, %matrixD2 = nvgpu.wargroup.mma %descA, %descB, %matrixC1, %matrixC2 : ....

// Epilogue 
nvgpu.wargroup.mma.store [%matrixD1, %matrixD2] to %sharedMemoryBuffer
  : !nvgpu.wargroup.accumulator<fragmented = vector<64x128xf32>>, 
    !nvgpu.wargroup.accumulator<fragmented = vector<64x128xf32>>
    into memref<128x128xf32,3>
```

3) One can do `nvgpu.wargroup.mma.store` without variadic as well: 

```
// Epilogue 
%s1 = memref.subview %sharedMemoryBuffer ....
nvgpu.wargroup.mma.store %matrixD1 to %s1
  : !nvgpu.wargroup.accumulator<fragmented = vector<64x128xf32>> into memref<128x128xf32,3>
%s2 = memref.subview %sharedMemoryBuffer ....
nvgpu.wargroup.mma.store %matrixD2 to %s2
  : !nvgpu.wargroup.accumulator<fragmented = vector<64x128xf32>> into memref<128x128xf32,3>
```

I have implemented **1st** Option. I guess you are advocating the **2nd** one. Is that right?

grypp replied:
I decided to simplify the `nvgpu.wargroup.accumulator` type. I am going to get rid of the varidic, but I need to change it for these 3 ops.

The IR will look like this instead of 1st option I wrote above

```
// Init
%matrixC1, %matrixC2 = nvgpu.wargroup.mma.init.accumulator ->  
                    !nvgpu.wargroup.accumulator<fragmented = vector<128x128xf32>>

// GEMM
%matrixD = nvgpu.wargroup.mma %descA, %descB, %matrixC ...

// Epilogue 
nvgpu.wargroup.mma.store [%matrixD1, %matrixD2] to %sharedMemoryBuffer
  : !nvgpu.wargroup.accumulator<fragmented = vector<128x128xf32>>
    into memref<128x128xf32,3>
```

grypp replied:
I've implemented #68728 (depends on this PR) that simplifies the IR. We get rid of variadic, and now the IR isn't dense.

qcolombet replied:
> I have implemented 1st Option. I guess you are advocating the 2nd one. Is that right?

Correct

qcolombet replied:
> I plan to simplify the `nvgpu.wargroup.accumulator` type by allowing larger types `<128x128>` instead of 2x`<64x128>`. The transformation will handle the creation of multiple structs if necessary. This change will enable me to manage everything with a single Op and eliminate the need for varidic in these three ops.
> 
> Here's how the updated IR will look (instead of option 1 above):
> 
> ```
> // Init
> %matrixC1, %matrixC2 = nvgpu.wargroup.mma.init.accumulator ->  
>                     !nvgpu.wargroup.accumulator<fragmented = vector<128x128xf32>>
> 
> // GEMM
> %matrixD = nvgpu.wargroup.mma %descA, %descB, %matrixC ...
> 
> // Epilogue 
> nvgpu.wargroup.mma.store [%matrixD1, %matrixD2] to %sharedMemoryBuffer
>   : !nvgpu.wargroup.accumulator<fragmented = vector<128x128xf32>>
>     into memref<128x128xf32,3>
> ```
> 
> I believe it's a good idea to proceed with this PR first, and then I will put the simplification PR later.

Works for me

qcolombet commented on mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td:
```diff
@@ -727,4 +727,15 @@ def NVGPU_WarpgroupMmaOp : NVGPU_Op<"warpgroup.mma"> {
+  }];
+  let results = (outs Variadic<NVGPU_WarpgroupAccumulator>:$matrixC);
+  let assemblyFormat = "attr-dict `->` type($matrixC)";
+}
```
Should we have a verifier to make sure we set a number of registers (e.g., reject size like `34x128xf32`) that makes sense?

grypp replied:
Let's do this incrementally. Because verification is different for different data type. I need to think verification holistically for all the nvgpu.warpgroup Ops

grypp commented on mlir/test/Conversion/NVGPUToNVVM/nvgpu-to-nvvm.mlir:
```diff
@@ -772,6 +772,77 @@ func.func @warpgroup_mma_128_128_64(
+  //CHECK: %[[S63:.+]] = llvm.insertvalue %[[S1]], %[[S62]][61] : !llvm.struct
+  //CHECK: %[[S64:.+]] = llvm.insertvalue %[[S1]], %[[S63]][62] : !llvm.struct
+  //CHECK: %[[S65:.+]] = llvm.insertvalue %[[S1]], %[[S64]][63] : !llvm.struct
+  %matrixC = nvgpu.warpgroup.mma.init.accumulator -> !nvgpu.warpgroup.accumulator< fragmented = vector<64x128xf32>>
```
Yes it needs to be in virtual registers. We generate a struct for that

I'm not sure I really understood the question.

grypp replied:
The accumulator (or the struct) is used by the `wgmma.mma_async` PTX instruction. You pass the members this struct as operands. 

Do you think there might be case that the struct can end up in different place?

qcolombet replied:
> Yes it needs to be in virtual registers. We generate a struct for that
> 
> I'm not sure I really understood the question.

You got the question right :).

I was afraid that this may produce some spill, but thinking more about it, even if that's the case, the uses in `wgmma.*` should force them back to register.

So nothing to worry about.

grypp replied:
Now you got me thinking :) It is actually possible for llvm to put the struct in local memory. I don't have a good answer on how to prevent this.

qcolombet replied:
I don't think you can :)

grypp committed:
[MLIR][NVGPU] Introduce `warpgroup.init.accumulator` Op #67530

This Op generates and initilizes the accumulator matrix for `nvgpu.warpgroup.mma` op to perform matrix-multiply-and-accumulate (mma).

Its associated transformation generates `!llvm.struct<>` and fill it with the initial values. The size of struct is number of required inout registers for `nvgpu.warpgroup.mma` op.

grypp committed:
address comments

grypp committed:
add verifier

qcolombet commented on mlir/include/mlir/Dialect/NVGPU/IR/NVGPU.td:
```diff
@@ -728,4 +728,16 @@ def NVGPU_WarpgroupMmaOp : NVGPU_Op<"warpgroup.mma"> {
+  let summary = "Initializes accumulator matrix for `warppgroup.mma`";
+
+  let description = [{
+    This Op generates and initilizes the accumulator matrix for 
```
The typo is still here :).
Did you forget to add this part of the change?

grypp replied:
Apologize. I am using vscode for github comments. Things are going wrong.

qcolombet reviewed (commented):
I'm still unsure about the variadic part of the new `init` instruction, as in, I don't understand what it buys us. Maybe a snippet of an example would help.

Let's do one more iteration on that, but other than that, lgtm.

grypp committed:
fix typo

qcolombet reviewed (approved):
LGTM
